lib/cusolver/multigpu.jl

```diff
@@ -0,0 +1,357 @@
+            flush(stdout)
+            throw(e)
+        end
+        return new(desc[])
+    end
+end
+
+Base.cconvert(::Type{cudaLibMgMatrixDesc_t}, obj::CudaLibMGDescriptor) = obj.desc
+
+mutable struct CudaLibMGGrid
+    desc::Ref{cudaLibMgGrid_t}
+
+    function CudaLibMGGrid(num_row_devs, num_col_devs, deviceIds, mapping)
+        desc = Ref{cudaLibMgGrid_t}()
+        cudaLibMgCreateDeviceGrid(desc, num_row_devs, num_col_devs, deviceIds, mapping)
+        return new(desc)
+    end
+end
+
+Base.cconvert(::Type{cudaLibMgGrid_t}, obj::CudaLibMGGrid) = obj.desc[]
+
+function allocateBuffers(n_row_devs, n_col_devs, descr, mat::Matrix)
+    mat_row_block_size = div(size(mat, 1), n_row_devs)
+    mat_col_block_size = div(size(mat, 2), n_col_devs)
+    mat_buffers  = Vector{CuMatrix}(undef, ndevices())
+    mat_numRows  = Vector{Int64}(undef, ndevices())
+    mat_numCols  = Vector{Int64}(undef, ndevices())
+    typesize = sizeof(eltype(mat))
+    ldas = Vector{Int64}(undef, ndevices())
+    current_dev = device()
+    mat_cpu_bufs = Vector{Matrix{eltype(mat)}}(undef, ndevices())
+    for (di, dev) in enumerate(devices())
+        ldas[di]    = mat_col_block_size
+        dev_row     = mod(di - 1, n_row_devs) + 1
+        dev_col     = div(di - 1, n_row_devs) + 1
+
+        mat_row_inds     = ((dev_row-1)*mat_row_block_size+1):min(dev_row*mat_row_block_size, size(mat, 1))
+        mat_col_inds     = ((dev_col-1)*mat_col_block_size+1):min(dev_col*mat_col_block_size, size(mat, 2))
+        mat_cpu_bufs[di] = Array(mat[mat_row_inds, mat_col_inds])
+    end
+    for (di, dev) in enumerate(devices())
+        device!(dev)
+        mat_gpu_buf = CuMatrix{eltype(mat)}(undef, size(mat))
+        unsafe_copyto!(mat_gpu_buf, 1, mat_cpu_bufs[di], 1, length(mat_cpu_bufs[di]))
+        mat_buffers[di] = mat_gpu_buf
+    end
+    for (di, dev) in enumerate(devices())
+        device!(dev)
+        synchronize()
+    end
+    device!(current_dev)
+    return mat_buffers
+end
+
+function returnBuffers(n_row_devs, n_col_devs, row_block_size, col_block_size, desc, dDs, D)
+    row_block_size = div(size(D, 1), n_row_devs)
+    col_block_size = div(size(D, 2), n_col_devs)
+    numRows  = [row_block_size for dev in 1:ndevices()]
+    numCols  = [col_block_size for dev in 1:ndevices()]
+    typesize = sizeof(eltype(D))
+    current_dev = device()
+    cpu_bufs = Vector{Matrix{eltype(D)}}(undef, ndevices())
+    for (di, dev) in enumerate(devices())
+        device!(dev)
+        dev_row = mod(di - 1, n_row_devs) + 1
+        dev_col = div(di - 1, n_row_devs) + 1
+        row_inds = ((dev_row-1)*row_block_size+1):min(dev_row*row_block_size, size(D, 1))
+        col_inds = ((dev_col-1)*col_block_size+1):min(dev_col*col_block_size, size(D, 2))
+        cpu_bufs[di] = Matrix{eltype(D)}(undef, length(row_inds), length(col_inds))
+        unsafe_copyto!(cpu_bufs[di], 1, dDs[di], 1, length(cpu_bufs[di]))
```

Comment:
if I change this to an ordinary copyto, stuff breaks; is it expected that the length of dDs[di] doesn't match length(cpu_bufs[di])?

Reply:
I think so yes

Reply:
Just checked to confirm, yeah the tests fail with `copyto!`

Reply:
Update: fixed this.